### DISCUSSION TITLE
Required NuGet packages? Getting unresolved external symbols

### DISCUSSION THREAD

**Discussion by fbrier** [cefsharp/CefSharp#3699] [General]
I created a packages.config file:
```xml
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="chromiumembeddedframework.runtime.win-x64" version="91.1.21"/>
  <package id="cef.sdk" version="91.1.21"/>
</packages>
```
The NuGet package "cef.redist.x64" appears to be the same as "chromiumembeddedframework.runtime.win-x64". Do I need both NuGet packages? My own project is based on cefsimple in /tests, but I am getting tons unresolved symbols when linking, which appear to be defined in the above libraries and DLLs (having done dumpbin).
```
libcef_dll_wrapper.lib(media_sink_device_info_callback_cpptoc.obj) : error LNK2001: unresolved external symbol cef_get_min_log_level
```
I don't understand Ninja, other than it is a generator (of what I do not know), but if I am just pulling in the above pre-built .lib and DLLs and linking with libcef_dll_wrapper.lib, libcef.lib, and chrome_elf.lib, shouldn't that be sufficient? I don't need to use cmake/Ninja, do I? I created the export lib(s) from the dll(s), although it would be great if those were included the cef.redist.x64 NuGet package.

**Comment by amaitland** [comment]
> The NuGet package "cef.redist.x64" appears to be the same as "chromiumembeddedframework.runtime.win-x64". Do I need both NuGet packages?

They are different. cef.redist is probably the better choice as it includes the locals and swiftshader. 

> My own project is based on cefsimple in /tests, but I am getting tons unresolved symbols when linking, which appear to be defined in the above libraries and DLLs

The packages provided by this project are designed for use with CefSharp. Whilst you are welcome to use them in your project,  help with compiling cefsimple is outside the scope of this project. 

The libcef_dll_wrapper is compiled for dynamic linking. 

> I created the export lib(s) from the dll(s), although it would be great if those were included the cef.redist.x64 NuGet package.

The cef.redist packages contain only the runtime. The cef.sdk package has the required lib files.

You can look at the vcxproj files within this project as a reference to see how the cef.sdk project is used.

https://github.com/cefsharp/CefSharp/blob/cefsharp/91/CefSharp.BrowserSubprocess.Core/CefSharp.BrowserSubprocess.Core.vcxproj

https://github.com/cefsharp/CefSharp/blob/cefsharp/91/CefSharp.BrowserSubprocess.Core/CefSharp.BrowserSubprocess.Core.vcxproj#L77
https://github.com/cefsharp/CefSharp/blob/cefsharp/91/CefSharp.BrowserSubprocess.Core/CefSharp.BrowserSubprocess.Core.vcxproj#L104
https://github.com/cefsharp/CefSharp/blob/cefsharp/91/CefSharp.BrowserSubprocess.Core/CefSharp.BrowserSubprocess.Core.vcxproj#L113

**Reply by fbrier** [reply]
Thank you for the response. I was planning on doing the bulk of my coding in C#. The CEF initialization and shutdown is getting called from a C++ plugin framework, so I don't have a choice in that. I was able to build the cef-binary project (assuming that is the one that creates the NuGet packages) using VS2019/v142 after adding the ARM64 to VS2019. Still trying to figure out why there are unresolved symbols. The goal at this point is to see whether all these tools/libraries will play nice with each other.

**Reply by amaitland** [reply]
> The CEF initialization and shutdown is getting called from a C++ plugin framework

What do you mean by this exactly? Is CEF already used by the host application? You cannot load multiple versions of CEF within a process.

**Comment by fbrier** [comment]
No CEF is not currently used or part of the system. Like other developers that have integrated CEF with Unreal Engine and Unity, I am trying to integrate it with the Godot game engine. I am trying to use their GDNative plugin architecture which has C++ entry points for initialization and shutdown.

Unfortunately, they have chosen yet another build system, SCONS, which (as far as I can tell) does not understand linking by just specifying a shared library with the .dll extension. It keeps trying to make it libcef.dll.lib. So I had to generate a .DEF and then a .LIB for the libcef.dll and chrome_elf.dll, and try and to link those, but am still getting unresolved symbols.

I just read through the 1000 line long build.ps1 looking for anything that might have been modifying the code (changing name mangling), but there was nothing. Long, but straightforward.

SCONS is nice, it's just CEF is using CMake/Ninja, Godot is using SCONS, Microsoft uses MSBuild. It WOULD be nice if we could all use the same build tool, but I get it. We are all looking to make our build chain easier to use, more efficient, and have different ideas on how to do it. Probably more content than you wanted, but thank you for the warning. I am sure there are more than a few statics and singletons in the CEF codebase to avoid doing that. Again, thank you for the help.

**Reply by amaitland** [reply]
> I am trying to integrate it with the Godot game engine. I am trying to use their GDNative plugin architecture which has C++ entry points for initialization and shutdown.

Better to discuss with whoever maintains Godot. Quick search suggests Godot uses mono for it's CLR implementation, it's unlikely that CefSharp will work as it's written using C++/CLI.

For reference CEF has it's own support forum at https://magpcss.org/ceforum/index.php

>  It keeps trying to make it libcef.dll.lib

There should be a lib file for libcef in the cef.sdk. There is no lib files for chrome_elf or any of the other libcef dependencies.

I'd suggest you use a standard distribution from https://cef-builds.spotifycdn.com/index.html#windows64

**Reply by amaitland** [reply]
You could potentially bootstrap the net core clr. You could then write q generic plugin that simply loads a .Net dll 

https://github.com/dotnet/docs/blob/main/docs/core/tutorials/netcore-hosting.md

**Reply by fbrier** [reply]
Your suggestion to pull the binary from the standard distribution combined with the CEF SDK Nuget resolved all the references, I am not sure what I was doing wrong building the .DEF and .LIB files from the DLL. I still would like the ceflib as a NuGet package. Perhaps I will fork the repo and add that target/artifact to the standard distribution build. Thank you for your help.
